src/lib.rs

```diff
@@ -1580,13 +1591,32 @@ pub fn run<F: Function>(
     options: &RegallocOptions,
 ) -> Result<Output, RegAllocError> {
     match options.algorithm {
-        Algorithm::Ion => ion::run(func, env, options.verbose_log, options.validate_ssa),
+        Algorithm::Ion => {
+            let mut ctx = Ctx::default();
+            run_with_ctx(func, env, options, &mut ctx)?;
+            Ok(ctx.output)
+        }
         Algorithm::Fastalloc => {
             fastalloc::run(func, env, options.verbose_log, options.validate_ssa)
         }
     }
 }
 
+/// Run the allocator.
+pub fn run_with_ctx<F: Function>(
+    func: &F,
+    env: &MachineEnv,
+    options: &RegallocOptions,
+    ctx: &mut Ctx,
+) -> Result<(), RegAllocError> {
```

Comment:
Could we return the `Output` here as with `run()`, by `mem::take`'ing it from the `Ctx`? I'd much prefer that to an implicit result in `ctx.output` -- OK to be imperative and implicit inside RA2 but a functional API is best.

Reply:
It is implicit because the allocations inside the output are reused, of course, If we assume the user will return the output to the `Ctx` this is fine but IMO I'd rather hint in the doc where the output is located, that way you don't need to write more code to get optimal performance.

Reply:
Actually returning a reference could be a good middle ground